software/contrib/egressus_melodiam.py

```diff
@@ -0,0 +1,910 @@
+
+                # Add time diff between clocks to inputClockDiffs Fifo list, skipping the first clock as we have no reference
+                if self.clockStep > 0:
+                    self.inputClockDiffs[self.clockStep % CLOCK_DIFF_BUFFER_LEN] = (
+                        newDiffBetweenClocks
+                    )
+
+                # Clock rate change detection
+                if (
+                    self.clockStep >= CLOCK_DIFF_BUFFER_LEN
+                    and abs(newDiffBetweenClocks - self.averageMsBetweenClocks)
+                    > MIN_CLOCK_CHANGE_DETECTION_MS
+                ):
+                    # Update average ms between clocks
+                    self.averageMsBetweenClocks = self.average(self.inputClockDiffs)
+                    # Clock rate or output division changed, recalculate optimal sample rate
+                    self.calculateOptimalSampleRate()
+
+                self.handleClockStep()
+                # Incremenent the clock step
+                self.clockStep += 1
+                self.newClockToProcess = False
+
+            # Cycle through outputs, process when needed
+            for idx in range(len(cvs)):
+                if (
+                    ticks_diff(ticks_ms(), self.lastSlewVoltageOutputTime[idx])
+                    >= self.msBetweenSamples[idx]
+                    and self.running
+                ):
+
+                    try:
+
+                        # Do we have a sample in the buffer?
+                        if self.slewBufferPosition[idx] < self.slewBufferSampleNum[idx]:
+                            # Yes, we have a sample, output voltage to match the sample, reset underrun counter and advance position in buffer
+
+                            # If a square interpolation mode (0) a precalculated value is used
+                            # as the generator function for square waves was really buggy
+                            if self.outputSlewModes[idx] == 0:
+                                v = self.squareOutputs[idx]
+                            else:
+                                v = next(self.slewGeneratorObjects[idx])
+                            cvs[idx].voltage(v)
+                            self.previousOutputVoltage[idx] = v
+                            self.bufferUnderrunCounter[idx] = 0
+                        else:
+                            # We do not have a sample - buffer under run
+                            # Output the previous voltage to keep things as smooth as possible
+                            cvs[idx].voltage(self.previousOutputVoltage[idx])
+                            self.bufferUnderrunCounter[idx] += 1
+
+                        # Advance the position in the sample/slew buffer
+                        self.slewBufferPosition[idx] += 1
+
+                        # Update the last sample output time
+                        self.lastSlewVoltageOutputTime[idx] = ticks_ms()
+                
+                    except StopIteration:

```

Comment:
[question]
What's actually raising the StopIteration exception that's being handled here?

Reply:
It happens when a generator function has yielded it's last value. It is not actually an exception in the true sense.
A generator function serving up it's last sample should be normal behaviour in the script - hence me adding it back. I should not have removed it earlier.

Reply:
Got it, thanks.